src/pubsub.ts

```diff
@@ -57,56 +59,65 @@ export class AMQPPubSub implements PubSubEngine {
-      .then(disposer => {
-        this.subsRefsMap[routingKey] = [
-          ...(this.subsRefsMap[routingKey] || []),
-          id,
-        ];
-        if (this.unsubscribeMap[routingKey]) {
-          return disposer();
-        }
-        this.unsubscribeMap[routingKey] = disposer;
-        return Promise.resolve(id);
-      });
+      return id;
     }
+
+    this.subsRefsMap[routingKey] = [
+      ...(this.subsRefsMap[routingKey] || []),
+      id,
+    ];
+
+    const existingDispose = this.unsubscribeMap[routingKey];
+    // Get rid of exisiting subscription while we get a new one.
+    const [newDispose] = await Promise.all([
+      this.subscriber.subscribe(this.exchange, routingKey, this.onMessage),
+      existingDispose ? existingDispose() : Promise.resolve(),
+    ]);
+
+    this.unsubscribeMap[routingKey] = newDispose;
+    return id;
   }
 
-  public unsubscribe(subId: number): Promise<void> {
-    const routingKey = this.subscriptionMap[subId].routingKey;
-    const refs = this.subsRefsMap[routingKey];
+  public async unsubscribe(subId: number): Promise<void> {
+    logger('Unsubscribe id: "%s"', subId);
+    const sub = this.subscriptionMap[subId];
+    if (!sub) {
+      throw new Error(`There is no subscription for id "${subId}"`);
+    }
+    const { routingKey } = sub;
 
+    const refs = this.subsRefsMap[routingKey];
     if (!refs) {
-      throw new Error(`There is no subscription of id "${subId}"`);
+      throw new Error(`There is no subscription ref for routing key "${routingKey}", id "${subId}"`);
     }
 
     if (refs.length === 1) {
       delete this.subscriptionMap[subId];
       return this.unsubscribeForKey(routingKey);
-    } else {
-      const index = refs.indexOf(subId);
-      const newRefs =
-        index === -1
-          ? refs
-          : [...refs.slice(0, index), ...refs.slice(index + 1)];
-      this.subsRefsMap[routingKey] = newRefs;
-      delete this.subscriptionMap[subId];
     }
-    return Promise.resolve();
+
+    const index = refs.indexOf(subId);
+    const newRefs =
+      index === -1
+        ? refs
+        : [...refs.slice(0, index), ...refs.slice(index + 1)];
+    this.subsRefsMap[routingKey] = newRefs;
+    delete this.subscriptionMap[subId];
   }
 
   public asyncIterator<T>(triggers: string | string[]): AsyncIterator<T> {
     return new PubSubAsyncIterator<T>(this, triggers);
   }
 
-  private onMessage(routingKey: string, message: any): void {
+  private onMessage = (routingKey: string, message: any): void => {
     const subscribers = this.subsRefsMap[routingKey];
 
-    // Don't work for nothing..
+    // Don't work for nothing...
     if (!subscribers || !subscribers.length) {
-      this.unsubscribeForKey(routingKey);
+      this.unsubscribeForKey(routingKey)
+        .catch((err) => {
+          logger('onMessage unsubscribeForKey error "%j", Routing Key "%s"', err, routingKey);
+        });
```

Comment:
please keep the catch on the same indent as the promise

Reply:
Will do.